### ISSUE TITLE
Cornerstone.imageCache image cache takes up to 2 times the memory

### ISSUE THREAD

**Issue by aiv367** [cornerstonejs/cornerstone#126]
Cornerstone.imageCache.cachedImages [n] .image;
Cornerstone.imageCache.imageCache.xxx.image;
An image cache will take 2 memory, 1.10.7 version is 0.9.0 version of the memory occupied by 2 times.
Easily lead to browser memory crash

**Comment by swederik** [comment]
Hi there,

I am a bit confused. In imageCache we are indeed adding the same cachedImage to both the `imageCacheDict` Object and `cachedImages` array, but that doesn't mean it will take up more memory. They are both referencing the same cachedImage.

Do you have any way to reproduce the issue? Is this happening with one specific image loader scheme?

**Comment by aiv367** [comment]
thank you for your reply.
After testing, really Cornerstone.imageCache.cachedImages [n] .image is equal to Cornerstone.imageCache.imageCache.xxx.image. Should this place not take up too much memory. But 1.10.7 version than 0.9.0 version, do take more than twice the memory, see cornerstone.js source I did not find the cause of the problem. Hope your help

**Comment by swederik** [comment]
It might be from this commit:

https://github.com/chafey/cornerstone/commit/5846cf0a5b6663c1bf0c9a7e145127b4abddd7fe

where we made the render canvas specific to each enabled element. This was required to fix usage on Safari.

**Comment by aiv367** [comment]
I wrote a small example, hoping to reproduce my question, please see
var imageIds = [
	'imageId1', 'imageId2', 'imageId3', ..., 'imageId1000'
];

imageIds.forEach (function (imageId) {
	cornerstone.loadAndCacheImage (imageId);
});

Before using the 0.13.3 version of the cornerstoneWADOImageLoader.js, the memory is normal, no problem. Later version upgrade, use 0.14.3 version cornerstoneWADOImageLoader.js,
Found that the memory occupancy situation is very different, almost more than doubled memory. This problem has been troubled me for a long time, has been studying your source, so far, did not find a specific cause of the problem, want to get your help, save me

**Comment by aiv367** [comment]
I found the reason for the problem.
Through the webworker.postMessage (data) to pass the past object, in cornerstoneWADOImageLoaderWebWorker.js received parameters (onmessage (msg)), not the address reference pass, but a separate copy. Worker data in the address because of the relationship, will not be released by the browser garbage collection mechanism, will still be retained. So it will cause 2 times the memory footprint.

**Comment by swederik** [comment]
Aha I see. That's interesting. I wonder what we can do to make sure it is GC'd. Thanks for looking into this!

**Comment by aiv367** [comment]
This problem after further investigation and found that the problem is not a webworker, I try to modify the cornerstoneWADOImageLoader.js code for testing and verification, after the end of each task are implemented webWorkers [msg.data.workerIndex] .worker.terminate (); and then create a new Worker object, assigned to go back and found that memory will still be 2 times the growth. I really is no solution, and back to the origin.

**Comment by kofifus** [comment]
I cannot reproduce this, can a test case be provided ?
Otherwise this should be closed.

**Comment by dannyrb** [comment]
Internally, we've noticed some memory issues with `cornerstoneWADOImageLoader`, but we've managed a band-aid fix that periodically kills idle workers, or restarts them after processing X images.

In either case, this is the wrong repo for this discussion. I'm going to close this issue. If someone can provide a good minimum reproduction and advice around how to measure/isolate/tackle this issue, then I would welcome a new issue over on the `cornerstoneWADOImageLoader` repository.
